src/hypofuzz/provider.py

```diff
@@ -26,29 +80,409 @@ def fresh_choice(
+        )
+        self.corpus = Corpus(self.db, self.database_key)
+
+        # Report that we've started this fuzz target
+        self.db.save(test_keys_key, self.database_key)
+        # save the worker identity once at startup
+        self.db.save_worker_identity(self.database_key, self.worker_identity)
+        # restore our saved minimal covering corpus, as well as any failures to
+        # replay.
+        for shrunk in [True, False]:
+            priority = (
+                ReplayPriority.FAILURE_SHRUNK
+                if shrunk
+                else ReplayPriority.FAILURE_UNSHRUNK
+            )
+            for choices in self.db.fetch_failures(self.database_key, shrunk=shrunk):
+                observation = self.db.fetch_failure_observation(
+                    self.database_key, choices
+                )
+                if observation is None:
+                    continue
+                self._enqueue(priority, choices, extra_data=observation)
+
+        # TODO: do we need/want this?
+        # self._enqueue(
+        #     ReplayPriority.COVERING,
+        #     (ChoiceTemplate(type="simplest", count=None),),
+        # )
+        for choices in self.db.fetch_corpus(self.database_key):
+            self._enqueue(ReplayPriority.COVERING, choices)
+
+        if not self._replay_queue:
+            # if no worker has ever worked on this test before, save an initial
+            # Phase.GENERATE report, so the graph displays a point at (0, 0).
+            #
+            # This is a hack. We should move this to the dashboard plotting code
+            # instead, by adding a virtual (0, 0) point.
+            self.phase = Phase.GENERATE
+            self._save_report(self._report)
+            self.phase = None
+
+        self._started = True
+
+    def _enqueue(
+        self,
+        priority: ReplayPriority,
+        choices: ChoicesT,
+        *,
+        extra_data: Optional[Any] = None,
+    ) -> None:
+        self._replay_queue.add((priority, choices, extra_data))
+
+    def on_event(self, event: DatabaseEvent) -> None:
+        # Some worker has found a new covering corpus element. Replay it in
+        # this worker.
+        if event.type == "save":
+            if event.key is DatabaseEventKey.CORPUS:
+                # the worker which saved this choice sequence might have been *us*,
+                # or we might simply already have this choice sequence in our
+                # corpus.
+                if (
+                    self.corpus is not None
+                    and Choices(event.value) in self.corpus.corpus
+                ):
+                    return
+                self._enqueue(ReplayPriority.COVERING, event.value)
+            # (should we also replay failures found by other workers? we may not
+            # want to stop early, since we could still find a different failure.
+            # but I guess the same logic would apply to the worker which found the
+            # failure..)
```

Comment:
Yes, we do want to enqueue failures found by other workers - but then _what to do with that_ is fairly complicated.  We don't want to spend much time duplicating shrinking when another worker is doing that; but likewise we don't want to abandon the test if that worker happens to drop out.

Let's open an issue to revisit this later.  Current proposal: add to queue, but also stop working on this test for a few minutes if the shrunk instance is smaller than this workers best known failing input (for that interesting origin), and then revisit.  Make the queue consumer smarter about dropping not-smallest failures instead of replaying them all.